File: AtCoder/ABC209/E.py
from collections import deque

N = int(input())
S = []
D = {}      # D['abc'] == 0
Dinv = []   # Dinv[0]  == 'abc'
Edges = []

SFST = set()

for _ in range(N):
  s = input()
  S.append(s)
  sf = s[:3]
  st = s[-3:]

  sfst = sf + st
  if sfst in SFST: continue
  SFST.add(sfst)

  if sf not in D:
    D[sf] = len(Dinv)
    Dinv.append(sf)
  
  if st not in D:
    D[st] = len(Dinv)
    Dinv.append(st)
  
  isf = D[sf]
  ist = D[st]
  Edges.append((isf, ist))

V = len(Dinv)
E = [[] for _ in range(V)]
Einv = [[] for _ in range(V)]
Deg = [0]*V # 出次数
for isf, ist in Edges:
  E[isf].append(ist)
  Einv[ist].append(isf)
  Deg[isf] += 1

queue = deque([i for i in range(V) if len(E[i]) == 0]) # 出次数0のノード
nodeTypes = [None]*V
#print(queue, [Dinv[q] for q in queue])

while queue:
  q = queue.popleft()

  nodeType = 'L'
  for e in E[q]:
    if nodeTypes[e] == 'L':
      nodeType = 'W'
      break
  nodeTypes[q] = nodeType

  for p in Einv[q]:
    Deg[p] -= 1
    if Deg[p] == 0 or (nodeType == 'L' and nodeTypes[p] is None):
      queue.append(p)
      Deg[p] = 0

for s in S:
  st = s[-3:]
  ist = D[st]

  nodeType = nodeTypes[ist]
  if   nodeType is None:
    print('Draw')
  elif nodeType == 'W':
    print('Aoki')
  else:
    print('Takahashi')

for key, value in D.items():
  print(key, value, nodeTypes[value])

File: AtCoder/ABC209/E.rs
static DRAW: usize = 0;
static LOSE: usize = 1;
static WIN: usize = 2;

fn read_line() -> String {
    let mut inp = String::new();
    std::io::stdin().read_line(&mut inp).unwrap();
    String::from(inp.trim())
}

fn main() {
    let n: usize = read_line().parse().unwrap();
    let m: usize = 52 * 52 * 52;

    let mut edges: Vec<Vec<usize>> = (0..m).map(|_| vec![]).collect();
    let mut words = vec![];
    let mut counts = vec![-1; m];
    for _ in 0..n {
        let s: Vec<_> = read_line().bytes().map(|c| if c >= b'a' { c - b'a' + 26 } else { c - b'A' }).collect();
        let len_s = s.len();

        let f0 = s[0] as usize;
        let f1 = s[1] as usize;
        let f2 = s[2] as usize;
        let f = f0 * 52 * 52 + f1 * 52 + f2;

        let t0 = s[len_s - 3] as usize;
        let t1 = s[len_s - 2] as usize;
        let t2 = s[len_s - 1] as usize;
        let t = t0 * 52 * 52 + t1 * 52 + t2;

        edges[t].push(f);
        words.push((f, t));
        counts[f] = 0;
        counts[t] = 0;
    }

    for &(f, _) in words.iter() {
        counts[f] += 1;
    }

    let queue: Vec<_> = counts.iter().enumerate().filter(|&(_, &c)| c == 0).map(|(i, _)| (i, LOSE)).collect();
    let mut queue = std::collections::VecDeque::from(queue);

    // println!("{:?}", counts.iter().enumerate().filter(|&(_, &c)| c != -1).collect::<Vec<_>>());
    // println!("{:?}", queue);

    let mut ans = vec![DRAW; m];

    while let Some((i, a)) = queue.pop_front() {
        if ans[i] != DRAW {
            continue;
        }
        ans[i] = a;
        // println!("{:?}", (i, a));

        if a == WIN {
            for &j in edges[i].iter() {
                counts[j] -= 1;
                if counts[j] == 0 {
                    queue.push_back((j, LOSE));
                }
            }
        }

        if a == LOSE {
            for &j in edges[i].iter() {
                queue.push_back((j, WIN));
            }
        }
    }

    for &(_, t) in words.iter() {
        if ans[t] == DRAW {
            println!("Draw");
        }
        if ans[t] == WIN {
            println!("Aoki");
        }
        if ans[t] == LOSE {
            println!("Takahashi");
        }
    }
}

// abc -> bcd -> cda
// ada -> ada
